sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStore.scala

```diff
@@ -1111,60 +1159,57 @@ object StateStore extends Logging {
     }
   }
 
+  // Wait until this partition can be processed
+  private def awaitProcessThisPartition(
+      id: StateStoreProviderId,
+      timeoutMs: Long
+  ): Boolean = maintenanceThreadPoolLock synchronized  {
+    val endTime = System.currentTimeMillis() + timeoutMs
+
+    // If immediate processing fails, wait with timeout
+    var canProcessThisPartition = processThisPartition(id)
+    while (!canProcessThisPartition && System.currentTimeMillis() < endTime) {
+      canProcessThisPartition = processThisPartition(id)
+      maintenanceThreadPoolLock.wait(timeoutMs)
+    }
+
+    canProcessThisPartition
+  }
+
+  private def doMaintenance(): Unit = doMaintenance(StateStoreConf.empty)
+
   /**
    * Execute background maintenance task in all the loaded store providers if they are still
    * the active instances according to the coordinator.
    */
-  private def doMaintenance(): Unit = {
+  private def doMaintenance(storeConf: StateStoreConf): Unit = {
     logDebug("Doing maintenance")
     if (SparkEnv.get == null) {
       throw new IllegalStateException("SparkEnv not active, cannot do maintenance on StateStores")
     }
+
+    // Providers that couldn't be processed now and need to be added back to the queue
+    val providersToRequeue = new ArrayBuffer[(StateStoreProviderId, StateStoreProvider)]()
+
+    while (!unloadedProvidersToClose.isEmpty) {
```

Comment:
Access to `unloadedProvidersToClose` is not truly thread safe here. Can we add a separate lock to guard access ?

Reply:
`unloadedProvidersToClose` is a concurrent queue, so we don't need lock here. Since this is a while loop, if a new elem is added, we will see it. Even if we don't, not an issue, we will see it on next maintenance